### DISCUSSION TITLE
[QUESTION] network splitting, last storage worker and recovery

### DISCUSSION THREAD

**Discussion by fulanya** [longhorn/longhorn#7809] [General]
## Question

I would like to discuss a scenario that I encountered in my environment where Longhorn is deployed across three zones with nodes labeled as storage=true. To simulate network splittings and  other issues, I followed these steps:

    Created Table A and disabled storage in Zone A (standby, everything works correctly with Longhorn).
    Created Table B and disabled storage in Zone B (standby, everything works correctly again with Longhorn).
    Created Table C and disabled storage in Zone C (standby, still functional).
    Attempted to create Table D, which failed because there was no associated storage.

Here's the interesting part: when I reactivated the node in Zone A (which was on standby) and subsequently activated the rest, I was able to recover Table A, but not Table B and the subsequent tables. However, if I had activated the node in Zone C first, I wouldn't have experienced any data loss.

I've encountered a scenario where it seems there might be a single correct way to proceed, and I'm curious if Longhorn has already implemented this or if it's feasible. The proposed behavior is as follows:

    1. Detection of Storage Issue: Once Longhorn detects a problem with the storage (when none is available), it should wait for a certain period until **the last available storage** becomes accessible again. During this time, the associated StatefulSet should continuously report errors.

    2. Flexibility with Timeouts: To provide more flexibility and achieve automatic recoveries, it could be beneficial to define a timeout period. Once this timeout is reached, Longhorn could attempt to mount any available storage, resolving the issue.

QUESTION: Is this currently possible with Longhorn, and if so, how can it be controlled? (auto/manual) Thank you in advance.

**Comment by c3y1huang** [comment]
Could you share the steps for reproducing the issue, including specific commands and simple workload manifests you've used? Additionally:
- Could you elaborate on what are the Tables?
- What was the process for putting a node into standby mode?
- How did you reactivate the nodes that are in standby mode?

**Comment by fulanya** [comment]
Updated Scenario:

There are three ESXi servers, each with three virtual machines:

ESXi Server-0:

- k8s-master0
- k8s-worker0
- k8s-storage0

ESXi Server-1:

- k8s-master1
- k8s-worker1
- k8s-storage1

ESXi Server-2:

- k8s-master2
- k8s-worker2
- k8s-storage2

**Additional Configuration:**

Longhorn is configured with defaultReplicaCount: 3. Storage in Longhorn is only available on nodes labeled as storage. A PostgreSQL database has been deployed on the k8s-worker1 node, using Longhorn as the storage class.

**Steps in the Scenario:**

**Table A and Storage Deactivation in Zone A:**
 

1. Table A was created in PostgreSQL.
2. Storage in Zone A was deactivated by putting the k8s-storage0 virtual machine in standby.
3. Result: Everything worked correctly with Longhorn, as there were still three operational instances of Longhorn.

**Table B and Storage Deactivation in Zone B:**

1. Table B was created in PostgreSQL.
2. Storage in Zone B was deactivated by putting the k8s-storage1 virtual machine in standby.
3. Result: Everything worked correctly with Longhorn, as there were still two operational instances of Longhorn.

**Table C and Storage Deactivation in Zone C:**

1. Table C was created in PostgreSQL.
2. Storage in Zone C was deactivated by putting the k8s-storage2 virtual machine in standby.
3. Result: Although storage was deactivated in Zone C, the functionality was still operational.

**Attempt to Create Table D without Associated Storage:**

An attempt was made to create Table D, but it failed because there was no associated available storage.

**Reactivation of Node in Zone A:**

By reactivating the node in Zone A (resuming/continuing the k8s-storage0 virtual machine that was in standby) and subsequently activating the other nodes:
**Result: Table A, which was previously present, remained accessible., but Table B and subsequent tables were not.** However, if the node in Zone C had been activated first, no data loss would have been experienced.

**Comment by fulanya** [comment]
It seems that I have partially found the solution, and it is done with the parameter **autoSalvage: false;** In this case, recovery is manual and can be done through the screen. In our case, we would select from storage 0, 1, and 2. The method this way would be "safe" (although it requires manual intervention). However, the only information I see associated with the storages is: a few seconds ago, a minute ago, etc. If all three zones "fail" in the same minute, they could lose hundreds of transactions that occurred in that minute. 

(1) How can we determine which storage (0, 1, 2) has the highest timestamp, in other words, the one that was last updated? 
(2)  Could you too confirm that the autoSalvage parameter does exactly what I have indicated? 
![image](https://github.com/longhorn/longhorn/assets/9947335/c01c348c-fc65-4df0-bd96-42de0be77519)

**Reply by c3y1huang** [reply]
> (1) How can we determine which storage (0, 1, 2) has the highest timestamp, in other words, the one that was last updated?

I am not sure about the behavior when having the ESXi virtual machines in standby mode and then reactivating it again in your scenario. You might want to explore the `lastHealthyAt` and `lastFailedAt` in the replica spec, or compare each node replica's revision-counters in `/var/lib/longhorn/replicas/<name>/revision.counter`.

> (2) Could you too confirm that the autoSalvage parameter does exactly what I have indicated

In theory, with auto-salvage disabled, you should be able to choose the replica to salvage from manually. I've not tried it, perhaps you can share your findings.

**Comment by fulanya** [comment]
First, I would like to apologize for the delay in response. Thank you for your observation in point 1; I will verify it when I have time. Regarding the second point, I want to inform you that if autosalvage=false, the system does not automatically repair itself in a scenario of unexpected node failure (or network splitting), as defined in Longhorn. On the other hand, during manual repair (refer to the previous image), you can choose which PVCs will be part of the repair. From what I have noticed, if all PVCs are selected, it seems to always be done correctly. In other words, the PVC with the latest information is always chosen.

Now, here is my improvement suggestion for Longhorn: 1) Longhorn should have a parameter like autoSalvageIfAllVolumesArePresent (or similar), in which it attempts automatic repair only if all the volumes defined (before the failure) are online and functioning correctly (see previous Image). In this case, it would be automatic and secure, meaning no information loss would occur. If all the defined volumes are not in the "present" state, a recheck should be performed, for example, every 1 minute, ...
